### ISSUE TITLE
Defaults vs custom values in attribute

### ISSUE THREAD

**Issue by migig** [stefanprodan/MvcThrottle#6]
I'm having trouble understanding the way defaults are used.

I always define my limits on the `[EnableThrottling(PerFoo=123)]` attribute. But there are also defaults set in the global filter, just to make the constructor happy.

I want the ones in my attributes to always take precedence. How do I do that? Sometimes they conflict.

Is it safe to do soemthing like this: 
    `Policy = new ThrottlePolicy(perSecond:long.MaxValue) { .... `

...and then set my real values on the attribute? I never want to rely on defaults.

**Comment by stefanprodan** [comment]
Yes setting the defaults to 0 should work, the default values supplied in the constructor applies to actions that don't have the EnableThrottling attribute, or when you've missed to supply a PerFoo rate in the attribute.
So if the default policy has a perDay limit, and the EnableThrottling doesn't have it, then the perDay limit from ThrottlePolicy constructor will be used.
If all your action have EnableThrottling will every PerX specified then none of the constructor values will apply.

**Comment by migig** [comment]
Okay that makes good sense.

But what happens, for example, if I want to throttle by day. In the attribute, is it enough to specify a value for `perDay`, or must I also specify for `perSecond` and `perHour` as well?

**Comment by migig** [comment]
"You need to put the rate types you want to enable in the constructor"... I assume you mean the ctor of the global filter. It's for a big project, so we "enabled" all of them to `0`.

Okay I tested and it seems to work, but we have many places where it's used, and I'm trying to go through many combinations to see what happens.

**Comment by migig** [comment]
It's a pity this isn't on the main docs, so for anybody else who wants to know, here is my understanding:

- The global filter takes rate limits. These are the defaults.
- If you don't specify custom rate limits in the `[EnableThrottling]` attribute, then the defaults are used.
- There are various ways to do throttling. If you prefer to specify custom limits on your attributes, and never rely on defaults, then you must set `0` as the default for all rate limits. So you do this: `Policy = new ThrottlePolicy(perSecond:0, perMinute:0, perHour:0, perDay:0, perWeek:0) { ...`. Then specify a rate limit in the attribute, for example: `[EnableThrottling(PerHour = 10)]`.
- If you take this approach, then the custom rate limits will override the defaults. But also, if your attribute is for e.g. `[EnableThrottling(PerHour = 10)]` then it not only overrides the `perHour:` default, but also ignores the `perSecond:` and `perMinute:` defaults. This means you don't need to specify custom values for all limits, *only the one you are interested in*. You don't need to do this: `[EnableThrottling(PerSecond = x, PerMinute = y, PerHour = 10)].

Please correct anything if it's wrong? I'll update more when I have tested more.

**Comment by stefanprodan** [comment]
You are right, this is the intended behavior. I must find a better way to enable/disable limit types. 
If don't specify a rate limit type when you register the filter, even if you specified that type in the attribute it will get ignored.

Thanks for the running these tests.

**Comment by migig** [comment]
Oops, so I misunderstood... :smile:

In your example I thought the `PerHour` will be used, but you say it will be ignored (because it has no corresponding default in the filter). So that means that there will be "maximum" throttling for that attribute, because `PerHour` is ignored, and everything else is `0`?

Sorry, I am maybe a bit "slow" (I've been throttled)...

I agree we should think what is a good way to handle this use case. Or keep it as it is, but make a "formal" documentation for it...
